Code/GraphMol/SynthonSpaceSearch/SynthonSpace.cpp

```diff
@@ -0,0 +1,719 @@
+    // be accumulated even if the molecule itself doesn't make it into the
+    // results set, for example if it isn't a random selection or it's
+    // outside maxHits or hitStart.
+    std::set<std::string> resultsNames;
+    // The random sampling, if requested, doesn't always produce the required
+    // number of hits in 1 sweep.  Also, totHits, being an upper bound, may
+    // not be achievable.
+    size_t tmpMaxHits(params.maxHits);
+    int numSweeps = 0;
+    while (results.size() < std::min(tmpMaxHits, totHits) && numSweeps < 10) {
+      buildHits(allHits, query, params, totHits, resultsNames, results);
+      // totHits is an upper bound, so may not be reached.
+      if (params.maxHits == -1 || !params.randomSample) {
+        break;
+      }
+      numSweeps++;
+    }
+  }
+  return SubstructureResults{std::move(results), totHits};
+}
+
+namespace {
+inline std::vector<std::string> splitLine(const std::string &str,
+                                          const std::regex &regexz) {
+  return {std::sregex_token_iterator(str.begin(), str.end(), regexz, -1),
+          std::sregex_token_iterator()};
+}
+
+// The Enamine/Chemspace files come with the connection points on the
+// synthons marked with [U], [Np], [Pu], [Am].  These need to be converted
+// to dummy atoms with isotope labels (1, 2, 3, 4 respectively) which can
+// be done safely on the SMILES string.
+void fixConnectors(std::string &smiles) {
+  const static std::vector<std::regex> connRegexs{
+      std::regex(R"(\[U\])"), std::regex(R"(\[Np\])"), std::regex(R"(\[Pu\])"),
+      std::regex(R"(\[Am\])")};
+  const static std::vector<std::string> repls{"[1*]", "[2*]", "[3*]", "[4*]"};
+  for (size_t i = 0; i < connRegexs.size(); ++i) {
+    smiles = std::regex_replace(smiles, connRegexs[i], repls[i]);
+  }
+}
+
+// Return a bitset giving the different connector types in this
+// molecule.
+boost::dynamic_bitset<> getConnectorPattern(
+    const std::vector<std::unique_ptr<ROMol>> &fragSet) {
+  boost::dynamic_bitset<> conns(4);
+  for (const auto &frag : fragSet) {
+    for (const auto &a : frag->atoms()) {
+      if (!a->getAtomicNum() && a->getIsotope()) {
+        conns.set(a->getIsotope() - 1);
+      }
+    }
+  }
+  return conns;
+}
+
+// Make pattern fps for the fragments and re-order fps and fragments into
+// descending order of number of bits set in the fp (the "largest fragment
+// heuristic").  The FP with the largest number of bits
+// is the most likely to screen out a matching synthon set since smaller,
+// less complex fragments are more likely to match something, so screen
+// with that first.
+std::vector<std::unique_ptr<ExplicitBitVect>> makePatternFPs(
+    std::vector<std::unique_ptr<ROMol>> &molFrags) {
+  std::vector<std::unique_ptr<ExplicitBitVect>> pattFPs;
+  for (const auto &frag : molFrags) {
+    pattFPs.emplace_back(
+        std::unique_ptr<ExplicitBitVect>(PatternFingerprintMol(*frag, 2048)));
+  }
```

Comment:
I think this works as:
```suggestion
    pattFPs.emplace_back(PatternFingerprintMol(*frag, 2048));
  }
```
`emplace_back` takes the arguments to the ctor of the contained object, so you can pass in a raw pointer.